File: js/src/tokenize/tokenize.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { tokenize } from "./tokenize.js";

// The tokenizer is proven against the shared fixture — every case must match the
// Python reference exactly. This is the EXEMPLAR every §4 algorithm test in this
// port follows: load conformance/cases/<algo>.json, assert deep equality over
// ALL cases, no leniency.
/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 11 };

describe("tokenize conformance", () => {
  const cases = loadCase<{ input: string; tokens: string[] }[]>("tokenize.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const c of cases) {
    it(`tokenize(${JSON.stringify(c.input)})`, () => {
      expect(tokenize(c.input)).toEqual(c.tokens);
    });
  }
});

File: golang/tokenize/tokenize_test.go
package tokenize

import (
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedTokenizeCases pins the size of conformance/cases/tokenize.json. A
// per-case loop proves nothing if the file silently shrinks, so the count is a
// binding part of the contract, not a "greater than zero" floor.
const expectedTokenizeCases = 11

// TestTokenizeVectorCount fails loudly if the fixture gains or loses vectors.
func TestTokenizeVectorCount(t *testing.T) {
	var cases []struct {
		Input  string   `json:"input"`
		Tokens []string `json:"tokens"`
	}
	conform.Case(t, "tokenize.json", &cases)
	if len(cases) != expectedTokenizeCases {
		t.Fatalf("tokenize.json: got %d cases, want %d", len(cases), expectedTokenizeCases)
	}
}

// The tokenizer is proven against the shared fixture — every case must match the
// Python reference exactly. This test is the EXEMPLAR every §4 algorithm test in
// this port follows: load conformance/cases/<algo>.json, assert byte-identical
// output over ALL cases, no leniency.
func TestTokenizeConformance(t *testing.T) {
	var cases []struct {
		Input  string   `json:"input"`
		Tokens []string `json:"tokens"`
	}
	conform.Case(t, "tokenize.json", &cases)

	if len(cases) != expectedTokenizeCases {
		t.Fatalf("tokenize.json: got %d cases, want %d", len(cases), expectedTokenizeCases)
	}
	for _, c := range cases {
		got := Tokenize(c.Input)
		want := c.Tokens
		if want == nil {
			want = []string{}
		}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("Tokenize(%q) = %v, want %v", c.Input, got, want)
		}
	}
}
